geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/iterators/AggregatingScan.scala

```diff
@@ -54,106 +53,164 @@ trait AggregatingScan[T <: AggregatingScan.Result]
+    val transformSft = options.get(TransformDefsOpt).map { t =>
+      val ts = options.getOrElse(TransformSchemaOpt,
+        throw new IllegalArgumentException("Defined a transform but no transform schema"))
+      val transformSft = IteratorCache.sft(ts)
+      val reusableTransformSf = TransformSimpleFeature(IteratorCache.sft(spec), transformSft, t)
       reusableTransformSf.setFeature(reusableSf)
+      aggregateSf = reusableTransformSf // note: side-effect in map
+      transformSft
     }
-    hasTransform = transform.isDefined
-
     val sampling = sample(options)
     val cql = options.get(CqlOpt).map(IteratorCache.filter(sft, spec, _))
     validate = (cql, sampling) match {
-      case (None, None)             => _ => true
-      case (Some(filt), None)       => filt.evaluate(_)
-      case (None, Some(samp))       => samp.apply
-      case (Some(filt), Some(samp)) => f => filt.evaluate(f) && samp.apply(f)
+      case (None, None)             => ValidateAll
+      case (Some(filt), None)       => new CqlValidator(filt)
+      case (None, Some(samp))       => new SampleValidator(samp)
+      case (Some(filt), Some(samp)) => new CqlSampleValidator(filt,  samp)
     }
-    result = initResult(sft, if (hasTransform) { Some(transformSft) } else { None }, options)
+    batchSize = options.get(BatchSizeOpt).map(_.toInt).getOrElse(defaultBatchSize)
+    result = initResult(sft, transformSft, batchSize, options)
   }
 
   /**
-    * Aggregates a batch of data. May not exhaust the underlying data
-    *
-    * @return encoded aggregate batch, or null if no results
-    */
-  def aggregate(): Array[Byte] = {
+   * Aggregates a batch of data. May not exhaust the underlying data
+   *
+   * @param callback callback to provide for results
+   * @return callback
+   */
+  def aggregate[A <: AggregateCallback](callback: A): A = {
     // noinspection LanguageFeature
     result.clear()
 
-    var rowValue = advance()
-    if (rowValue == null) {
-      // skip the result.isEmpty check if we haven't read any data
-      // FIXME there are some aggregates that are never empty
-      return null
+    val status = new AggregateStatus(callback)
+
+    var rowValue = try { if (hasNextData) { nextData() } else { null } } catch {
+      case NonFatal(e) => logger.error("Error in underlying scan while aggregating value:", e); null
     }
 
     while (rowValue != null) {
       try {
         reusableSf.setIdBuffer(rowValue.row, rowValue.rowOffset, rowValue.rowLength)
         reusableSf.setBuffer(rowValue.value, rowValue.valueOffset, rowValue.valueLength)
-        if (validateFeature(reusableSf)) {
+        if (validate(reusableSf)) {
           // write the record to our aggregated results
-          if (hasTransform) {
-            aggregateResult(reusableTransformSf, result)
-          } else {
-            aggregateResult(reusableSf, result)
-          }
+          status.aggregated(aggregateResult(aggregateSf, result))
```

Comment:
How does 'aggregateSf' get updated?

Reply:
aggregateSf is going to be either reusableSf (on line 56) or transformSf (on line 65). either way, it gets updated through reusableSf, as transformSf is a view on top of reusableSf. clear as mud?

Reply:
it's slightly convoluted b/c we run the filter against the non-transformed feature, then apply the transform in the result

Reply:
Yeah, got it the second or third time I read through the class.:)  I'm good now.